### ISSUE TITLE
~real64 instead of +real64

### ISSUE THREAD

**Issue by AlexisPerry** [spack/spack#4863]
`spack install xsdk` producess the following error:
```
==> Error: Invalid spec: 'metis@5.1.0%gcc@5.4.0~debug~gdb~int64~real64+shared arch=linux-centos7-x86_64 ^cmake@3.7.1%gcc@5.4.0~doc+ncurses+openssl+ownlibs~qt arch=linux-centos7-x86_64'. Package metis requires variant +real64, but spec asked for ~real64
```
I looked at the metis package and can't figure out where the '~real64' is coming from.

**Comment by davydden** [comment]
This is a concretizer bug. U can get around it by being more explicit in xsdk w.r.t. PETSc dependency (see dealii package for an example of this bugfix). Unfortunately u will still hit some other issues after this.

There was a discussion on the commit, but I can't find it now.

**Comment by adamjstewart** [comment]
@alalazo This is another reason we no longer forward variants in `hdf5`. @tgamblin Really need this concretizer fix. We have 51 open issues related to concretization.

**Comment by davydden** [comment]
FYI, that's how I made it go in dealii https://github.com/LLNL/spack/pull/4735

**Comment by alalazo** [comment]
This has been solved in the meanwhile (1d06949306b592c0f675dd959303ee1eda1a2cf2):
```console
$ spack spec xsdk
Input spec
--------------------------------
xsdk

Concretized
--------------------------------
xsdk@0.5.0%gcc@9.0.1+butterflypack~cuda+dealii~debug+ginkgo+libensemble+omega-h+phist+precice+strumpack+trilinos arch=linux-ubuntu18.04-broadwell
    ^alquimia@xsdk-0.5.0%gcc@9.0.1 build_type=RelWithDebInfo +shared arch=linux-ubuntu18.04-broadwell
        ^cmake@3.15.5%gcc@9.0.1~doc+ncurses+openssl+ownlibs patches=3387faf4a71efe81c0fa17410b270ca7d352081ac88d2322df3da9bb6a6a3f2d ~qt arch=linux-ubuntu18.04-broadwell
            ^ncurses@6.1%gcc@9.0.1~symlinks~termlib arch=linux-ubuntu18.04-broadwell
                ^pkgconf@1.6.3%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
            ^openssl@1.1.1d%gcc@9.0.1+systemcerts arch=linux-ubuntu18.04-broadwell
                ^perl@5.30.0%gcc@9.0.1+cpanm+shared+threads arch=linux-ubuntu18.04-broadwell
                    ^gdbm@1.18.1%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
                        ^readline@8.0%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
                ^zlib@1.2.11%gcc@9.0.1+optimize+pic+shared arch=linux-ubuntu18.04-broadwell
        ^hdf5@1.10.5%gcc@9.0.1~cxx~debug+fortran+hl+mpi patches=b61e2f058964ad85be6ee5ecea10080bf79e73f83ff88d1fa4b602d00209da9c +pic+shared~szip~threadsafe arch=linux-ubuntu18.04-broadwell
            ^numactl@2.0.12%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
                ^autoconf@2.69%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
                    ^m4@1.4.18%gcc@9.0.1 patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 +sigsegv arch=linux-ubuntu18.04-broadwell
                        ^libsigsegv@2.12%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
                ^automake@1.16.1%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
                ^libtool@2.4.6%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
            ^openmpi@3.1.4%gcc@9.0.1~cuda+cxx_exceptions fabrics=none ~java~legacylaunchers~memchecker~pmi schedulers=none ~sqlite3~thread_multiple+vt arch=linux-ubuntu18.04-broadwell
                ^hwloc@1.11.11%gcc@9.0.1~cairo~cuda~gl+libxml2~nvml+pci+shared arch=linux-ubuntu18.04-broadwell
                    ^libpciaccess@0.13.5%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
                        ^util-macros@1.19.1%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
                    ^libxml2@2.9.9%gcc@9.0.1~python arch=linux-ubuntu18.04-broadwell
                        ^libiconv@1.16%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
                        ^xz@5.2.4%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
        ^pflotran@xsdk-0.5.0%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
            ^petsc@3.12.1%gcc@9.0.1~X~batch clanguage=C ~complex~debug+double~fftw+hdf5+hypre~int64~knl+metis+mpi~mumps+shared~suite-sparse+superlu-dist+trilinos arch=linux-ubuntu18.04-broadwell
                ^hypre@2.18.2%gcc@9.0.1~complex~debug~int64~internal-superlu~mixedint+mpi~openmp+shared+superlu-dist arch=linux-ubuntu18.04-broadwell
                    ^openblas@0.3.7%gcc@9.0.1+avx2~avx512 cpu_target=auto ~ilp64+pic+shared threads=none ~virtual_machine arch=linux-ubuntu18.04-broadwell
                    ^superlu-dist@6.1.1%gcc@9.0.1 build_type=RelWithDebInfo ~int64~openmp+shared arch=linux-ubuntu18.04-broadwell
                        ^metis@5.1.0%gcc@9.0.1 build_type=Release ~gdb~int64 patches=4991da938c1d3a1d3dea78e49bbebecba00273f98df2a656e38b83d55b281da1,b1225da886605ea558db7ac08dd8054742ea5afe5ed61ad4d0fe7a495b1270d2 ~real64+shared arch=linux-ubuntu18.04-broadwell
                        ^parmetis@4.0.3%gcc@9.0.1 build_type=RelWithDebInfo ~gdb patches=4f892531eb0a807eb1b82e683a416d3e35154a455274cf9b162fb02054d11a5b,50ed2081bc939269689789942067c58b3e522c269269a430d5d34c00edbc5870,704b84f7c7444d4372cb59cca6e1209df4ef3b033bc4ee3cf50f369bce972a9d +shared arch=linux-ubuntu18.04-broadwell
                ^python@3.7.4%gcc@9.0.1+bz2+ctypes+dbm+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4~uuid+zlib arch=linux-ubuntu18.04-broadwell
                    ^bzip2@1.0.8%gcc@9.0.1+shared arch=linux-ubuntu18.04-broadwell
                        ^diffutils@3.7%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
                    ^expat@2.2.9%gcc@9.0.1+libbsd arch=linux-ubuntu18.04-broadwell
                        ^libbsd@0.10.0%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
                    ^gettext@0.20.1%gcc@9.0.1+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu18.04-broadwell
                        ^tar@1.32%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
                    ^libffi@3.2.1%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
                    ^sqlite@3.30.1%gcc@9.0.1~column_metadata+fts~functions~rtree arch=linux-ubuntu18.04-broadwell
                ^trilinos@12.18.1%gcc@9.0.1~adios2~alloptpkgs+amesos+amesos2+anasazi+aztec+belos+boost build_type=RelWithDebInfo ~cgns~chaco~complex~debug+dtk+epetra+epetraext~exodus+explicit_template_instantiation~float+fortran~fortrilinos+gtest+hdf5+hypre+ifpack+ifpack2~intrepid+intrepid2~isorropia+kokkos+metis~minitensor+ml+muelu~mumps+nox~openmp~phalanx~piro~pnetcdf~python~rol~rythmos+sacado+shards+shared~shylu~stk~suite-sparse~superlu+superlu-dist~teko~tempus+teuchos+tpetra~x11~xsdkflags~zlib+zoltan+zoltan2 arch=linux-ubuntu18.04-broadwell
                    ^boost@1.70.0%gcc@9.0.1+atomic+chrono~clanglibcpp~context~coroutine cxxstd=98 +date_time~debug+exception~fiber+filesystem+graph~icu+iostreams+locale+log+math~mpi+multithreaded~numpy~pic+program_options~python+random+regex+serialization+shared+signals~singlethreaded+system~taggedlayout+test+thread+timer~versionedlayout visibility=hidden +wave arch=linux-ubuntu18.04-broadwell
                    ^glm@0.9.7.1%gcc@9.0.1 build_type=RelWithDebInfo arch=linux-ubuntu18.04-broadwell
                    ^matio@1.5.13%gcc@9.0.1+hdf5+shared+zlib arch=linux-ubuntu18.04-broadwell
                    ^netcdf-c@4.7.2%gcc@9.0.1~dap~hdf4 maxdims=1024 maxvars=8192 +mpi~parallel-netcdf patches=56532470875b9a97f3cf2a7d9ed16ef1612df3265ee38880c109428322ff3a40,f52db13c61b9c19aafe03c2a865163b540e9f6dee36e3a5f808f05fac59f2030 +pic+shared arch=linux-ubuntu18.04-broadwell
    ^amrex@19.08%gcc@9.0.1~amrdata build_type=Release dimensions=3 ~eb~fortran+linear_solvers+mpi~openmp~particles precision=double ~shared~sundials arch=linux-ubuntu18.04-broadwell
    ^butterflypack@1.1.0%gcc@9.0.1 build_type=RelWithDebInfo +shared arch=linux-ubuntu18.04-broadwell
        ^arpack-ng@3.7.0%gcc@9.0.1+mpi+shared arch=linux-ubuntu18.04-broadwell
        ^netlib-scalapack@2.0.2%gcc@9.0.1 build_type=RelWithDebInfo patches=22ebf4e3d5a6356cd6086ea65bfdf30f9d0a2038136127590cd269d15bdb03af,e8f30dd1f26e523dfb552f8d7b8ad26ac88fc0c8d72e3d4f9a9717a3383e0b33 ~pic+shared arch=linux-ubuntu18.04-broadwell
    ^dealii@9.1.1%gcc@9.0.1+adol-c+arpack~assimp build_type=DebugRelease ~cuda cuda_arch=none ~doc~ginkgo~gmsh+gsl+hdf5~int64+metis+mpi+muparser+nanoflann~netcdf+oce~optflags+p4est+petsc~python+scalapack+slepc~sundials~symengine+threads+trilinos arch=linux-ubuntu18.04-broadwell
        ^adol-c@develop%gcc@9.0.1~advanced_branching+atrig_erf~boost+doc+examples~openmp~sparse arch=linux-ubuntu18.04-broadwell
        ^gsl@2.5%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
        ^intel-tbb@2019.8%gcc@9.0.1 cxxstd=default patches=ce1fb16fb932ce86a82ca87cf0431d1a8c83652af9f552b264213b2ff2945d73 +shared+tm arch=linux-ubuntu18.04-broadwell
        ^muparser@2.2.6.1%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
        ^nanoflann@1.2.3%gcc@9.0.1 build_type=RelWithDebInfo arch=linux-ubuntu18.04-broadwell
        ^oce@0.18.3%gcc@9.0.1~X11+tbb arch=linux-ubuntu18.04-broadwell
        ^p4est@2.2%gcc@9.0.1+mpi~openmp arch=linux-ubuntu18.04-broadwell
        ^slepc@3.12.0%gcc@9.0.1+arpack~blopex arch=linux-ubuntu18.04-broadwell
        ^suite-sparse@5.3.0%gcc@9.0.1~cuda~openmp+pic~tbb arch=linux-ubuntu18.04-broadwell
    ^ginkgo@1.1.0%gcc@9.0.1 build_type=Release ~cuda cuda_arch=none ~develtools~full_optimizations~openmp+shared arch=linux-ubuntu18.04-broadwell
    ^mfem@4.0.1-xsdk%gcc@9.0.1~conduit~debug+examples~gnutls+gzstream+hypre~lapack~libunwind+metis+miniapps~mpfr+mpi~netcdf~occa~openmp+petsc~pumi~raja~shared+static~suite-sparse+sundials+superlu-dist~threadsafe timer=auto arch=linux-ubuntu18.04-broadwell
        ^sundials@5.0.0%gcc@9.0.1+ARKODE+CVODE+CVODES+IDA+IDAS+KINSOL build_type=RelWithDebInfo ~cuda+examples-c~examples-cuda~examples-cxx~examples-f2003+examples-f77~examples-f90+examples-install~f2003~fcmix+generic-math+hypre~int64~klu~lapack+mpi~openmp patches=5d1dba283a14e136f07b8740f63310fe73d17c923e223b6380c11e888d9f5434 +petsc precision=double ~pthread~raja+shared+static+superlu-dist~superlu-mt arch=linux-ubuntu18.04-broadwell
    ^omega-h@9.29.0%gcc@9.0.1 build_type=RelWithDebInfo ~examples+mpi+optimize+shared+symbols~throw+trilinos~warnings+zlib arch=linux-ubuntu18.04-broadwell
    ^phist@1.8.0%gcc@9.0.1 build_type=RelWithDebInfo ~fortran~host kernel_lib=tpetra +mpi~openmp outlev=2 ~parmetis~scamac+shared~trilinos arch=linux-ubuntu18.04-broadwell
    ^plasma@19.8.1%gcc@9.0.1 build_type=RelWithDebInfo ~lua patches=6ce78aaea37672a3cd63cbf3cc5f9f12da974a099b8312c6b0678b4b55020287,97606686359b620b7eae22a769793f269dafd52d838bb519ef57d7cb60bea3de +shared arch=linux-ubuntu18.04-broadwell
    ^precice@1.6.1%gcc@9.0.1 build_type=RelWithDebInfo +mpi+petsc~python+shared arch=linux-ubuntu18.04-broadwell
        ^eigen@3.3.7%gcc@9.0.1 build_type=RelWithDebInfo ~fftw~metis~mpfr~scotch~suitesparse arch=linux-ubuntu18.04-broadwell
    ^pumi@2.2.1%gcc@9.0.1 build_type=RelWithDebInfo ~fortran~int64~shared simmodsuite=none ~zoltan arch=linux-ubuntu18.04-broadwell
    ^py-libensemble@0.5.2%gcc@9.0.1~mpi~nlopt+petsc4py~scipy arch=linux-ubuntu18.04-broadwell
        ^py-numpy@1.17.4%gcc@9.0.1+blas+lapack arch=linux-ubuntu18.04-broadwell
            ^py-setuptools@41.4.0%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
        ^py-petsc4py@3.12.0%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
            ^py-mpi4py@3.0.3%gcc@9.0.1 arch=linux-ubuntu18.04-broadwell
    ^strumpack@3.3.0%gcc@9.0.1~build_dev_tests~build_tests build_type=RelWithDebInfo +butterflypack+c_interface~count_flops+mpi+openmp~parmetis~scotch~shared~task_timers arch=linux-ubuntu18.04-broadwell
    ^tasmanian@7.0%gcc@9.0.1+blas build_type=Release ~cuda~fortran~magma+mpi+openmp patches=e4ccfb63933a684e7cba7783e09188af30023597c95dd2374fb3e882add5b205,f9e0a08f3f6ce916c8133096996acb7d5df699d12ed540707c8b1bc4481e5a53 ~python+xsdkflags arch=linux-ubuntu18.04-broadwell
```
Feel free to reopen if I missed anything.
